File: src/nodetool/models/prediction.py
from datetime import datetime
from typing import Any

from nodetool.models.base_model import DBField, DBIndex, DBModel, create_time_ordered_uuid
from nodetool.models.condition_builder import Field

"""
Defines the Prediction database model.

Represents a prediction or job execution within the nodetool system.
Stores details about the execution, including the user, workflow (optional),
node involved, provider, model used, status, timing, cost, and any logs or errors.

This model also serves for cost tracking and usage analytics for AI provider calls.
"""


@DBIndex(columns=["user_id", "provider"], name="idx_prediction_user_provider")
@DBIndex(columns=["user_id", "model"], name="idx_prediction_user_model")
@DBIndex(columns=["created_at"], name="idx_prediction_created_at")
class Prediction(DBModel):
    """Database model representing a prediction or job execution."""

    @classmethod
    def get_table_schema(cls):
        """Returns the database table schema for predictions."""
        return {
            "table_name": "nodetool_predictions",
        }

    id: str = DBField(hash_key=True)
    user_id: str = DBField()
    node_id: str = DBField(default="")
    provider: str = DBField(default="")
    model: str = DBField()
    workflow_id: str | None = DBField(default=None)
    error: str | None = DBField(default=None)
    logs: str | None = DBField(default=None)
    status: str = DBField(default="starting")
    created_at: datetime | None = DBField(default_factory=datetime.now)
    started_at: datetime | None = DBField(default=None)
    completed_at: datetime | None = DBField(default=None)
    cost: float | None = DBField(default=None)
    duration: float | None = DBField(default=None)
    hardware: str | None = DBField(default=None)

    # Token-based usage (for text/chat/embedding models)
    input_tokens: int | None = DBField(default=None)
    output_tokens: int | None = DBField(default=None)
    total_tokens: int | None = DBField(default=None)
    cached_tokens: int | None = DBField(default=None)
    reasoning_tokens: int | None = DBField(default=None)

    # Size-based usage (for image/audio/video models)
    input_size: int | None = DBField(default=None)  # Input data size in bytes
    output_size: int | None = DBField(default=None)  # Output data size in bytes

    # Model parameters (resolution, quality, voice, etc.)
    parameters: dict[str, Any] | None = DBField(default=None)

    # Additional metadata
    metadata: dict[str, Any] | None = DBField(default=None)

    @classmethod
    async def create(
        cls,
        **kwargs,
    ):
        """Creates a new prediction record in the database.

        Args:
            user_id: The ID of the user initiating the prediction.
            node_id: The ID of the node performing the prediction.
            provider: The name of the prediction provider (e.g., 'openai', 'anthropic').
            model: The specific model used for the prediction.
            workflow_id: Optional ID of the workflow this prediction belongs to.
            status: Initial status of the prediction (default: 'starting').
            cost: Optional estimated or actual cost of the prediction.
            duration: Optional duration of the prediction in seconds.
            hardware: Optional identifier for the hardware used.
            created_at: Optional timestamp when the prediction was created.
            started_at: Optional timestamp when the prediction started.
            completed_at: Optional timestamp when the prediction completed.
            input_tokens: Optional number of input/prompt tokens used.
            output_tokens: Optional number of output/completion tokens used.
            total_tokens: Optional total number of tokens used.
            cached_tokens: Optional number of cached tokens (for providers that support it).
            reasoning_tokens: Optional number of reasoning tokens (for providers that support it).
            input_size: Optional input data size in bytes (for image/audio/video models).
            output_size: Optional output data size in bytes (for image/audio/video models).
            parameters: Optional model-specific parameters (resolution, quality, voice, etc.).
            metadata: Optional additional metadata about the prediction.

        Returns:
            The newly created and saved Prediction instance.
        """
        created_at = kwargs.get('created_at')
        if created_at is None:
            created_at = datetime.now()
            kwargs['created_at'] = created_at
        
        if 'id' not in kwargs:
            kwargs['id'] = create_time_ordered_uuid()

        prediction = cls(**kwargs)
        await prediction.save()
        return prediction

    @classmethod
    async def find(cls, user_id: str, id: str):
        """Finds a prediction by its ID, ensuring it belongs to the specified user.

        Args:
            user_id: The ID of the user who should own the prediction.
            id: The ID of the prediction to find.

        Returns:
            The Prediction object if found and owned by the user, otherwise None.
        """
        prediction = await cls.get(id)
        if prediction is None or prediction.user_id != user_id:
            return None
        return prediction

    @classmethod
    async def paginate(
        cls,
        user_id: str,
        workflow_id: str | None = None,
        provider: str | None = None,
        model: str | None = None,
        limit: int = 100,
        start_key: str | None = None,
        reverse: bool = False,
    ):
        """Paginates through predictions for a user with optional filtering.

        Args:
            user_id: The ID of the user whose predictions to fetch.
            workflow_id: Optional workflow ID to filter predictions by.
            provider: Optional provider name to filter predictions by.
            model: Optional model name to filter predictions by.
            limit: Maximum number of predictions to return.
            start_key: The ID of the prediction to start pagination after (exclusive).
            reverse: Whether to return results in reverse chronological order.

        Returns:
            A tuple containing a list of Prediction objects and the ID of the
            last evaluated prediction (or an empty string if it's the last page).
        """
        # Build condition
        condition = Field("user_id").equals(user_id).and_(Field("id").greater_than(start_key or ""))

        if workflow_id:
            condition = condition.and_(Field("workflow_id").equals(workflow_id))
        if provider:
            condition = condition.and_(Field("provider").equals(provider))
        if model:
            condition = condition.and_(Field("model").equals(model))

        return await cls.query(
            condition=condition,
            limit=limit,
            reverse=reverse,
        )

    @classmethod
    async def aggregate_by_user(
        cls,
        user_id: str,
        provider: str | None = None,
        model: str | None = None,
    ) -> dict[str, Any]:
        """Aggregate cost and usage statistics for a user.

        Args:
            user_id: User ID to aggregate for
            provider: Optional provider filter
            model: Optional model filter

        Returns:
            Dictionary with aggregated totals

        Note:
            For production use with large datasets, consider implementing
            database-level aggregation queries instead of in-memory aggregation.
        """
        # Fetch records for the user with filters
        predictions, _ = await cls.paginate(
            user_id=user_id,
            provider=provider,
            model=model,
            limit=10000,  # High limit for aggregation
            reverse=False,
        )

        total_cost = 0
        total_input_tokens = 0
        total_output_tokens = 0
        total_tokens = 0
        for p in predictions:
            total_cost += p.cost or 0
            total_input_tokens += p.input_tokens or 0
            total_output_tokens += p.output_tokens or 0
            total_tokens += p.total_tokens or 0
        # If total_tokens wasn't tracked, calculate from input + output
        if total_tokens == 0:
            total_tokens = total_input_tokens + total_output_tokens
        call_count = len(predictions)

        return {
            "user_id": user_id,
            "provider": provider,
            "model": model,
            "total_cost": total_cost,
            "total_input_tokens": total_input_tokens,
            "total_output_tokens": total_output_tokens,
            "total_tokens": total_tokens,
            "call_count": call_count,
        }

    @classmethod
    async def aggregate_by_provider(
        cls,
        user_id: str,
    ) -> list[dict[str, Any]]:
        """Aggregate cost and usage by provider for a user.

        Args:
            user_id: User ID to aggregate for

        Returns:
            List of aggregations, one per provider

        Note:
            For production use with large datasets, consider implementing
            database-level GROUP BY queries instead of in-memory aggregation.
        """
        # Fetch all records for the user
        predictions, _ = await cls.paginate(
            user_id=user_id,
            limit=10000,  # High limit for aggregation
            reverse=False,
        )

        # Group by provider
        provider_stats: dict[str, dict[str, Any]] = {}
        for pred in predictions:
            if pred.provider not in provider_stats:
                provider_stats[pred.provider] = {
                    "provider": pred.provider,
                    "total_cost": 0.0,
                    "total_input_tokens": 0,
                    "total_output_tokens": 0,
                    "total_tokens": 0,
                    "call_count": 0,
                }

            stats = provider_stats[pred.provider]
            stats["total_cost"] += pred.cost or 0
            stats["total_input_tokens"] += pred.input_tokens or 0
            stats["total_output_tokens"] += pred.output_tokens or 0
            stats["total_tokens"] += pred.total_tokens or (pred.input_tokens or 0) + (pred.output_tokens or 0)
            stats["call_count"] += 1

        return list(provider_stats.values())

    @classmethod
    async def aggregate_by_model(
        cls,
        user_id: str,
        provider: str | None = None,
    ) -> list[dict[str, Any]]:
        """Aggregate cost and usage by model for a user.

        Args:
            user_id: User ID to aggregate for
            provider: Optional provider filter

        Returns:
            List of aggregations, one per model

        Note:
            For production use with large datasets, consider implementing
            database-level GROUP BY queries instead of in-memory aggregation.
        """
        # Fetch all records for the user
        predictions, _ = await cls.paginate(
            user_id=user_id,
            provider=provider,
            limit=10000,  # High limit for aggregation
            reverse=False,
        )

        # Group by model using tuple key to avoid collisions
        model_stats: dict[tuple[str, str], dict[str, Any]] = {}
        for pred in predictions:
            key = (pred.provider, pred.model)
            if key not in model_stats:
                model_stats[key] = {
                    "provider": pred.provider,
                    "model": pred.model,
                    "total_cost": 0.0,
                    "total_input_tokens": 0,
                    "total_output_tokens": 0,
                    "total_tokens": 0,
                    "call_count": 0,
                }

            stats = model_stats[key]
            stats["total_cost"] += pred.cost or 0
            stats["total_input_tokens"] += pred.input_tokens or 0
            stats["total_output_tokens"] += pred.output_tokens or 0
            stats["total_tokens"] += pred.total_tokens or (pred.input_tokens or 0) + (pred.output_tokens or 0)
            stats["call_count"] += 1

        return list(model_stats.values())

File: ts/packages/models/src/prediction.ts
/**
 * Prediction model -- tracks AI provider call costs and token usage.
 *
 * Port of Python's `nodetool.models.prediction`.
 */

import type { TableSchema } from "./database-adapter.js";
import type { Row } from "./database-adapter.js";
import {
  DBModel,
  createTimeOrderedUuid,
  type IndexSpec,
  type ModelClass,
} from "./base-model.js";
import { field } from "./condition-builder.js";

// ── Aggregate result types ──────────────────────────────────────────

export interface AggregateResult {
  user_id: string;
  provider?: string | null;
  model?: string | null;
  total_cost: number;
  total_input_tokens: number;
  total_output_tokens: number;
  total_tokens: number;
  call_count: number;
}

export interface ProviderAggregateResult {
  provider: string;
  total_cost: number;
  total_input_tokens: number;
  total_output_tokens: number;
  total_tokens: number;
  call_count: number;
}

export interface ModelAggregateResult {
  provider: string;
  model: string;
  total_cost: number;
  total_input_tokens: number;
  total_output_tokens: number;
  total_tokens: number;
  call_count: number;
}

// ── Schema ───────────────────────────────────────────────────────────

const PREDICTION_SCHEMA: TableSchema = {
  table_name: "nodetool_predictions",
  primary_key: "id",
  columns: {
    id: { type: "string" },
    user_id: { type: "string" },
    node_id: { type: "string" },
    provider: { type: "string" },
    model: { type: "string" },
    workflow_id: { type: "string", optional: true },
    cost: { type: "number", optional: true },
    input_tokens: { type: "number", optional: true },
    output_tokens: { type: "number", optional: true },
    total_tokens: { type: "number", optional: true },
    cached_tokens: { type: "number", optional: true },
    reasoning_tokens: { type: "number", optional: true },
    created_at: { type: "string" },
    metadata: { type: "string", optional: true },
  },
};

const PREDICTION_INDEXES: IndexSpec[] = [
  { name: "idx_predictions_user_id", columns: ["user_id"], unique: false },
  {
    name: "idx_predictions_user_provider",
    columns: ["user_id", "provider"],
    unique: false,
  },
];

// ── Model ────────────────────────────────────────────────────────────

export class Prediction extends DBModel {
  static override schema = PREDICTION_SCHEMA;
  static override indexes = PREDICTION_INDEXES;

  declare id: string;
  declare user_id: string;
  declare node_id: string;
  declare provider: string;
  declare model: string;
  declare workflow_id: string | null;
  declare cost: number | null;
  declare input_tokens: number | null;
  declare output_tokens: number | null;
  declare total_tokens: number | null;
  declare cached_tokens: number | null;
  declare reasoning_tokens: number | null;
  declare created_at: string;
  declare metadata: string | null;

  constructor(data: Row) {
    super(data);
    const now = new Date().toISOString();
    this.id ??= createTimeOrderedUuid();
    this.created_at ??= now;
    this.node_id ??= "";
    this.provider ??= "";
    this.model ??= "";
    this.workflow_id ??= null;
    this.cost ??= null;
    this.input_tokens ??= null;
    this.output_tokens ??= null;
    this.total_tokens ??= null;
    this.cached_tokens ??= null;
    this.reasoning_tokens ??= null;
    this.metadata ??= null;
  }

  // ── Static queries ───────────────────────────────────────────────

  static async paginate(
    userId: string,
    opts: {
      provider?: string | null;
      model?: string | null;
      limit?: number;
      startKey?: string;
    } = {},
  ): Promise<[Prediction[], string]> {
    const { limit = 50, provider, model } = opts;
    let cond = field("user_id").equals(userId);
    if (provider) cond = cond.and(field("provider").equals(provider));
    if (model) cond = cond.and(field("model").equals(model));

    return (Prediction as unknown as ModelClass<Prediction>).query({
      condition: cond,
      orderBy: "created_at",
      reverse: true,
      limit,
    });
  }

  static async aggregateByUser(
    userId: string,
    opts?: { provider?: string | null; model?: string | null },
  ): Promise<AggregateResult> {
    let cond = field("user_id").equals(userId);
    if (opts?.provider) cond = cond.and(field("provider").equals(opts.provider));
    if (opts?.model) cond = cond.and(field("model").equals(opts.model));

    const [predictions] = await (
      Prediction as unknown as ModelClass<Prediction>
    ).query({ condition: cond, limit: 10000 });

    let total_cost = 0;
    let total_input_tokens = 0;
    let total_output_tokens = 0;
    let total_tokens = 0;

    for (const p of predictions) {
      total_cost += p.cost ?? 0;
      total_input_tokens += p.input_tokens ?? 0;
      total_output_tokens += p.output_tokens ?? 0;
      total_tokens += p.total_tokens ?? 0;
    }

    return {
      user_id: userId,
      provider: opts?.provider ?? null,
      model: opts?.model ?? null,
      total_cost,
      total_input_tokens,
      total_output_tokens,
      total_tokens,
      call_count: predictions.length,
    };
  }

  static async aggregateByProvider(
    userId: string,
  ): Promise<ProviderAggregateResult[]> {
    const cond = field("user_id").equals(userId);
    const [predictions] = await (
      Prediction as unknown as ModelClass<Prediction>
    ).query({ condition: cond, limit: 10000 });

    const groups = new Map<string, ProviderAggregateResult>();
    for (const p of predictions) {
      let entry = groups.get(p.provider);
      if (!entry) {
        entry = {
          provider: p.provider,
          total_cost: 0,
          total_input_tokens: 0,
          total_output_tokens: 0,
          total_tokens: 0,
          call_count: 0,
        };
        groups.set(p.provider, entry);
      }
      entry.total_cost += p.cost ?? 0;
      entry.total_input_tokens += p.input_tokens ?? 0;
      entry.total_output_tokens += p.output_tokens ?? 0;
      entry.total_tokens += p.total_tokens ?? 0;
      entry.call_count += 1;
    }

    return [...groups.values()];
  }

  static async aggregateByModel(
    userId: string,
    opts?: { provider?: string | null },
  ): Promise<ModelAggregateResult[]> {
    let cond = field("user_id").equals(userId);
    if (opts?.provider) cond = cond.and(field("provider").equals(opts.provider));

    const [predictions] = await (
      Prediction as unknown as ModelClass<Prediction>
    ).query({ condition: cond, limit: 10000 });

    const groups = new Map<string, ModelAggregateResult>();
    for (const p of predictions) {
      const key = `${p.provider}::${p.model}`;
      let entry = groups.get(key);
      if (!entry) {
        entry = {
          provider: p.provider,
          model: p.model,
          total_cost: 0,
          total_input_tokens: 0,
          total_output_tokens: 0,
          total_tokens: 0,
          call_count: 0,
        };
        groups.set(key, entry);
      }
      entry.total_cost += p.cost ?? 0;
      entry.total_input_tokens += p.input_tokens ?? 0;
      entry.total_output_tokens += p.output_tokens ?? 0;
      entry.total_tokens += p.total_tokens ?? 0;
      entry.call_count += 1;
    }

    return [...groups.values()];
  }
}
